### PR TITLE
add test for embedded readme

### PR DESCRIPTION
**Pull request by lyndaidaii**
need upload readme file
address issue: https://github.com/NuGet/Engineering/issues/3659
EndToEnd test: https://nuget.visualstudio.com/NuGetBuild/_build/results?buildId=79899&view=results

### CODE REVIEW

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            var expectedPath = GetExpectedReadmePath(package);
+
+            // Act & Assert
+            await _clients.FlatContainer.WaitForPackageAsync(package.Id, package.NormalizedVersion, _logger);
+            var readmeContents = await _clients.FlatContainer.TryAndGetFileStringContent(package.Id, package.NormalizedVersion, FlatContainerContentType.Readme, _logger);
+            Assert.All(readmeContents, x => Assert.Equal(expectedContent, x));
+        }
+
+        [Fact]
+        public async Task PushedPackageWithReadmeIsAvailableInSearch()
```
Comment (skofman1) [inline src/NuGet.Services.EndToEnd/ReadmeTests.cs:54]: >PushedPackageWithReadmeIsAvailableInSearch [](start = 26, length = 42)

The readme link doesn't exist in Search. Why is this test needed?

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        [Fact]
+        public async Task PushedPackageWithReadmeIsAvailableInRegistration()
+        {
+            // Arrange
+            var package = await _pushedPackages.PrepareAsync(PackageType.EmbeddedReadmeFile, _logger);
+            var expectedPath = GetExpectedReadmePath(package);
+
+            // Act & Assert
+            var packageRegistrationList = await _clients.Registration.WaitForPackageAsync(package.Id, package.FullVersion, semVer2: false, logger: _logger);
```
Comment (skofman1) [inline src/NuGet.Services.EndToEnd/ReadmeTests.cs:35]: >false [](start = 132, length = 5)

we should test both SemVer 1.0 and SemVer 2.0 registration hives

File: src/NuGet.Services.EndToEnd/Support/TestData.cs [hunk 3]
```diff
@@ -78,15 +91,15 @@ public static Stream BuildPackageStream(PackageCreationContext context)
...
-                return InjectIconMetadata(packageStream, context);
-            }
+                return InjectIconOrReadmeMetadata(packageStream, context);
+            }
 
             return packageStream;
         }
 
-        private static Stream InjectIconMetadata(Stream packageStream, PackageCreationContext context)
+        private static Stream InjectIconOrReadmeMetadata(Stream packageStream, PackageCreationContext context)
```
Comment (skofman1) [inline src/NuGet.Services.EndToEnd/Support/TestData.cs:102]: >InjectIconOrReadmeMetadata [](start = 30, length = 26)

InjectEmbeddedMetadata?

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs (hunk 1 above)
Comment (lyndaidaii) [reply src/NuGet.Services.EndToEnd/ReadmeTests.cs:54]: Just want check if package is available in the search, I guess is not necessary here

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs (hunk 2 above)
Comment (loic-sharma) [reply src/NuGet.Services.EndToEnd/ReadmeTests.cs]: It looks like setting the `semVer2` parameter to `false` is the right thing to do actually. A value of `true` excludes the legacy hives. See:

https://github.com/NuGet/NuGet.Services.EndToEnd/blob/68f53566e117a99bde9c92ed882e8d4c712a02dc/src/NuGet.Services.EndToEnd/Support/Clients/RegistrationClient.cs#L136

https://github.com/NuGet/NuGet.Services.EndToEnd/blob/68f53566e117a99bde9c92ed882e8d4c712a02dc/src/NuGet.Services.EndToEnd/Support/Clients/V3IndexClient.cs#L51-L55

https://github.com/NuGet/NuGet.Services.EndToEnd/blob/68f53566e117a99bde9c92ed882e8d4c712a02dc/src/NuGet.Services.EndToEnd/Support/Clients/V3IndexClient.cs#L87-L92

Comment (loic-sharma) [reply src/NuGet.Services.EndToEnd/ReadmeTests.cs]: FYI @lyndaidaii, we have multiple copies of the registration API:

1. The legacy hive - This used by older clients that don't support modern features
1. The gzipped hive - This is for NuGet client versions 3.4.0 to 3.6.0. The JSON is compressed to improve performance.
1. The semver2 hive - This is for NuGet client versions 3.6.0 and newer. This includes SemVer 2.0.0 packages (package with versions like `1.0.0-prerelease+metadata`) that aren't supported by older clients.

We want to make sure that your package exists in all of these hives. If your package had a semver2 version, you wouldn't want to wait on the legacy and gzipped hives as the package weill never show up there. For more information on this, see our docs: https://docs.microsoft.com/en-us/nuget/api/registration-base-url-resource

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            Assert.All(packageRegistrationList, x => Assert.EndsWith(expectedPath, x.CatalogEntry.ReadmeUrl));
+        }
+
+        [Fact]
+        public async Task PushedPackageWithReadmeIsAvailableInFlatContainer()
+        {
+            // Arrange
+            var package = await _pushedPackages.PrepareAsync(PackageType.EmbeddedReadmeFile, _logger);
+            var expectedContent = GetReadmeData();
+            var expectedPath = GetExpectedReadmePath(package);
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/ReadmeTests.cs:47]: It looks like this is unused

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+            // Act & Assert
+            await _clients.FlatContainer.WaitForPackageAsync(package.Id, package.NormalizedVersion, _logger);
+            var readmeContents = await _clients.FlatContainer.TryAndGetFileStringContent(package.Id, package.NormalizedVersion, FlatContainerContentType.Readme, _logger);
+            Assert.All(readmeContents, x => Assert.Equal(expectedContent, x));
+        }
+
+        private static string GetExpectedReadmePath(Package package)
+        {
+            return $"/{package.Id}/{package.NormalizedVersion}#show-readme-container";
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/ReadmeTests.cs:57]: Would it be possible to use the `GalleryBaseUrl` config to build the full readme URL?

https://github.com/NuGet/NuGet.Services.EndToEnd/blob/68f53566e117a99bde9c92ed882e8d4c712a02dc/src/NuGet.Services.EndToEnd/Config/Dev.json#L7

File: src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj [hunk 6]
```diff
@@ -1,181 +1,183 @@
...
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="App.config">
+      <SubType>Designer</SubType>
+    </None>
+    <EmbeddedResource Include="Support\TestData\Icons\icon.png" />
+    <EmbeddedResource Include="Support\TestData\Icons\icon.jpg" />
+    <EmbeddedResource Include="Support\TestData\Readmes\readme.md" />
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj:116]: What are the whitespace changes in this file? Should we undo them?

File: src/NuGet.Services.EndToEnd/Support/Utilities/TestDataResourceUtility.cs [hunk 7]
```diff
@@ -1,38 +1,46 @@
...
+            }
+        }
+
+        public static string GetResourceStringContent(string name)
+        {
+            using (var reader = new StreamReader(GetManifestResourceStream(name)))
+            {
+                return reader.ReadToEnd();
+            }
+        }
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Utilities/TestDataResourceUtility.cs:29]: What are the whitespace changes in this file? Should we undo them?

File: test/NuGet.Services.EndToEnd.Test/Support/TestDataTests.cs [hunk 8]
```diff
@@ -1,58 +1,102 @@
...
+
+                using (var nuspecStream = nuspecEntry.Open())
+                using (var streamReader = new StreamReader(nuspecStream))
+                {
+                    var nuspecContent = streamReader.ReadToEnd();
+                    Assert.Contains($"<icon>{iconFilename}</icon>", nuspecContent);
+                }
+            }
+        }
+
```
Comment (loic-sharma) [inline test/NuGet.Services.EndToEnd.Test/Support/TestDataTests.cs:57]: What are the whitespace changes in this file? Should we undo them?

**Review by loic-sharma — APPROVED** [review]
Changes look good! It looks like there are some unintentional whitespace changes, please revert that if necessary.

File: src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj (hunk 6 above)
Comment (lyndaidaii) [reply src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj:116]: Sure, it was defualt format in VS setting. I will change them back

**Comment by zhhyu** [comment]
Looking now...

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+            // Act & Assert
+            await _clients.FlatContainer.WaitForPackageAsync(package.Id, package.NormalizedVersion, _logger);
+            var readmeContents = await _clients.FlatContainer.TryAndGetFileStringContent(package.Id, package.NormalizedVersion, FlatContainerContentType.Readme, _logger);
+            Assert.All(readmeContents, x => Assert.Equal(expectedContent, x));
+        }
+
+        private static string GetReadmeData()
+        {
+            var sourceFilename = $"Readmes.readme.md";
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/ReadmeTests.cs:55]:
```suggestion
            var sourceFilename = "Readmes.readme.md";

```

File: src/NuGet.Services.EndToEnd/Support/TestData/Readmes/readme.md [hunk 10]
```diff
@@ -0,0 +1 @@
+This is embedded readme test
```
Comment (zhhyu) [inline src/NuGet.Services.EndToEnd/Support/TestData/Readmes/readme.md:1]: nit: This is an embedded readme test

File: src/NuGet.Services.EndToEnd/Support/Clients/V2V3SearchClient.cs [hunk 11]
```diff
@@ -476,6 +476,7 @@ public class V3SearchPackage
             public List<V3VersionEntry> Versions { get; set; }
             public string LicenseUrl { get; set; }
             public string IconUrl { get; set; }
+            public string ReadmeUrl { get; set; }
```
Comment (zhhyu) [inline src/NuGet.Services.EndToEnd/Support/Clients/V2V3SearchClient.cs:479]: Could you please confirm whether we need this anywhere for readme tests?

File: test/NuGet.Services.EndToEnd.Test/Support/TestDataTests.cs [hunk 12]
```diff
@@ -1,58 +1,102 @@
...
+                using (var readmeDataStream = readmeEntry.Open())
+                using (var streamReader = new StreamReader(readmeDataStream))
+                {
+                    actualReadmeData = streamReader.ReadToEnd();
+                }
+
+                Assert.Equal(readmeData, actualReadmeData);
+
+                var nuspecEntry = zipArchive.GetEntry($"testP.nuspec");
```
Comment (zhhyu) [inline test/NuGet.Services.EndToEnd.Test/Support/TestDataTests.cs:90]: There is no need to use "string interpolation" here.

**Review by zhhyu — APPROVED** [review]

**Comment by zhhyu** [comment]
Please verify the "semVer2" change with other reviewers before checking in.

File: src/NuGet.Services.EndToEnd/Support/Clients/V2V3SearchClient.cs (hunk 11 above)
Comment (lyndaidaii) [reply src/NuGet.Services.EndToEnd/Support/Clients/V2V3SearchClient.cs:479]: good catch, I forgot to delete.

File: src/NuGet.Services.EndToEnd/ReadmeTests.cs (hunk 2 above)
Comment (lyndaidaii) [reply src/NuGet.Services.EndToEnd/ReadmeTests.cs:36]: I decided to to rename value here since if this is little confusing

**Comment by lyndaidaii** [comment]
E2E INT passed: https://nuget.visualstudio.com/NuGetBuild/_build/results?buildId=80520&view=results

**Pull request merged**